### PR TITLE
New module for managing NetApp E-Series iSCSI Interfaces

### PR DESCRIPTION
**Pull request by lmprice**
Define a new module for configuring NetApp E-Series iSCSI interfaces.

##### SUMMARY

Introduce a new module for managing E-Series iSCSI interfaces. Each E-Series system that supports iSCSI will have between 2 and 6 iSCSI interfaces (channels), that can be configured on each controller.

##### ISSUE TYPE
 - New Module Pull Request
##### COMPONENT NAME
netapp_e_iscsi_interface

##### ANSIBLE VERSION

```
ansible 2.6.0 (detached HEAD ce8c336d52) last updated 2018/04/18 11:23:03 (GMT -500)
  config file = None
  configured module search path = [u'/home/lorenp/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/lorenp/ansible/lib/ansible
  executable location = /home/lorenp/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```
changed: [testhost] => (item=1) => {"changed": true, "enabled": true, "item": 1, "msg": "The interface settings have been updated."}
```

### CODE REVIEW

File: test/integration/integration_config.yml [hunk 1]
```diff
@@ -1,3 +1,7 @@
 ---
 win_output_dir: 'C:\ansible_testing'
 output_dir: ~/ansible_testing
+netapp_e_ssid: 1
+netapp_e_api_host: 10.113.1.111:8443
+netapp_e_api_username: admin
+netapp_e_api_password: infiniti
```
Comment (mattclay) [inline test/integration/integration_config.yml]: Integration tests should not add anything to the shared integration config.

**Review by mattclay — DISMISSED** [review]

**Comment by lmprice** [comment]
@mattclay That issue should be resolved. Not sure how I can make the request resolved.

File: lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+            self.url += '/'
+
+        if self.mtu < 1500 or self.mtu > 9000:
+            self.module.fail_json(msg="The provided mtu is invalid, it must be > 1500 and < 9000 bytes.")
+
+    @property
+    def interfaces(self):
+        ifaces = list()
+        try:
+            (rc, ifaces) = request(self.url + 'storage-systems/%s/graph/xpath-filter?query=/controller/hostInterfaces'
```
Comment (hulquest) [inline lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:248]: This looks like a newer endpoint.  Does the documentation need to have a minimum version of SANtricity OS?

**Review by hulquest — COMMENTED** [review]
Once the question about backward compatibility is answered, this looks good to me.

File: lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py (hunk 2 above)
Comment (lmprice) [reply lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:248]: @hulquest We added this back in the WSP 1.3 time-frame (early 2016). While it's probably been in there long enough it doesn't need to be documented (IMO), that does beg the question of how to do so for when it does. I can add a note to the module if that makes sense.

Comment (hulquest) [reply lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:248]: I think a brief sentence is worthy.  Nothing is worse than trying something only to figure out the target API on the server isn't compatible.  We should try to remember to ask this question of all new modules and substantial changes.

Comment (lmprice) [reply lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:248]: @hulquest I added a note in the documentation. In the next patch I submit I'll see about adding a more lengthy blurb in the doc_fragments about API requirements, etc.

**Comment by lmprice** [comment]
@hulquest @amit0701 @schmots1 @carchi8py Can you give me a review on this patch?

**Review by carchi8py — APPROVED** [review]

**Comment by hulquest** [comment]
shipit

**Comment by lmprice** [comment]
@hulquest Just a theory, but can you hit review and mark it as "approved"?

**Review by ndswartz — APPROVED** [review]
shipit

**Review by carchi8py — APPROVED** [review]

File: lib/ansible/utils/module_docs_fragments/netapp.py [hunk 3]
```diff
@@ -136,7 +136,8 @@ class ModuleDocFragment(object):
     required: false
```
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/netapp.py]: Please remove this file from the PR

File: test/sanity/validate-modules/ignore.txt [hunk 4]
```diff
@@ -1133,7 +1133,6 @@ lib/ansible/modules/storage/netapp/na_cdot_volume.py E322
 lib/ansible/modules/storage/netapp/na_cdot_volume.py E324
 lib/ansible/modules/storage/netapp/na_cdot_volume.py E325
 lib/ansible/modules/storage/netapp/netapp_e_amg.py E322
-lib/ansible/modules/storage/netapp/netapp_e_amg.py E325
 lib/ansible/modules/storage/netapp/netapp_e_amg_role.py E322
```
Comment (gundalow) [inline test/sanity/validate-modules/ignore.txt]: Please remove this file from the PR

File: lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+            - When absent, the IPv4 configuration will be cleared and IPv4 connectivity disabled.
+        choices:
+            - present
+            - absent
+        default: present
+    address:
+        description:
+            - The IPv4 address to assign to the interface.
+            - Should be specified in xx.xx.xx.xx form.
+            - Mutually exclusive with I(config_method=dhcp)
```
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:58]: Add code to check for this and do `fail_json` if needed just after `argument_spec`

File: lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+    subnet_mask:
+        description:
+            - The subnet mask to utilize for the interface.
+            - Should be specified in xx.xx.xx.xx form.
+            - Mutually exclusive with I(config_method=dhcp)
```
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:63]: Add code to check for this and do `fail_json` if needed just after `argument_spec`

File: lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+    gateway:
+        description:
+            - The IPv4 gateway address to utilize for the interface.
+            - Should be specified in xx.xx.xx.xx form.
+            - Mutually exclusive with I(config_method=dhcp)
```
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:68]: Add code to check for this and do `fail_json` if needed just after `argument_spec`

File: lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+    config_method:
+        description:
+            - The configuration method type to use for this interface.
+            - dhcp is mutually exclusive with I(address), I(subnet_mask), and I(gateway).
```
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:72]: Add code to check for this and do `fail_json` if needed just after `argument_spec`

File: lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py (hunk 2 above)
Comment (gundalow) [reply lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:248]: @hulquest @lmprice Is this discussion resolved?

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py (hunk 2 above)
Comment (lmprice) [reply lib/ansible/modules/storage/netapp/netapp_e_iscsi_interface.py:248]: @gundalow Yes, I believe that it has. I added a comment in the notes about what API version is required for this module.

**Comment by lmprice** [comment]
@gundalow The issues should be resolved.

**Review by gundalow — APPROVED** [review]

**Pull request merged**